File: crates/autd3-rs-pattern-holo/src/linear_synthesis/gspat.rs
use core::num::NonZeroUsize;

use autd3_rs_core::common::Length;
use autd3_rs_core::geometry::Geometry;
use autd3_rs_core::value::{Intensity, Phase};

use crate::amplitude_target::AmplitudeTarget;
use crate::backend::LinAlgBackend;
use crate::constraint::IntensityConstraint;
use crate::directivity::Directivity;
use crate::error::HoloError;
use crate::linear_synthesis::batch::{BatchSetup, solve_batched};
use crate::mask::TransducerMask;
use crate::propagation::{make_propagation_matrix, quantize, target_amplitudes};

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct GspatOption<'a> {
    pub repeat: NonZeroUsize,
    pub constraint: IntensityConstraint,
    pub directivity: Directivity,
    pub mask: TransducerMask<'a>,
    pub parallel: bool,
}

impl Default for GspatOption<'_> {
    fn default() -> Self {
        Self {
            repeat: NonZeroUsize::new(100).unwrap(),
            constraint: IntensityConstraint::Clamp(Intensity::MIN, Intensity::MAX),
            directivity: Directivity::Sphere,
            mask: TransducerMask::AllEnabled,
            parallel: true,
        }
    }
}

#[allow(clippy::many_single_char_names)]
pub fn gspat<B: LinAlgBackend>(
    backend: &B,
    geometry: &Geometry,
    foci: &[AmplitudeTarget],
    wavelength: Length,
    option: &GspatOption<'_>,
    phases: &mut [Vec<Phase>],
    intensities: &mut [Vec<Intensity>],
) -> Result<(), HoloError> {
    if foci.is_empty() {
        return Err(HoloError::NoFoci);
    }
    crate::mask::validate_dst_len(phases.len(), geometry)?;
    crate::mask::validate_dst_len(intensities.len(), geometry)?;
    let mask = option.mask;
    mask.validate(geometry)?;

    let g = make_propagation_matrix(
        backend,
        geometry,
        foci,
        wavelength,
        option.directivity,
        mask,
    );
    let b = backend.back_prop(&g);
    let amps = target_amplitudes(backend, foci);

    let r = backend.gemm(&g, &b);

    let mut gamma =
        backend.repeat_gemv_normalized(&r, backend.gemv(&r, &amps), &amps, option.repeat.get() - 1);
    backend.amplitude_correct(&mut gamma, &amps);
    let q = backend.gemv(&b, &gamma);

    quantize(
        backend,
        geometry,
        &q,
        option.constraint,
        mask,
        option.parallel,
        phases,
        intensities,
    );
    Ok(())
}

pub fn gspat_batch<B: LinAlgBackend>(
    backend: &B,
    geometry: &Geometry,
    foci: &[AmplitudeTarget],
    wavelength: Length,
    option: &GspatOption<'_>,
    phases: &mut [Vec<Vec<Phase>>],
    intensities: &mut [Vec<Vec<Intensity>>],
) -> Result<(), HoloError> {
    let setup = BatchSetup {
        constraint: option.constraint,
        directivity: option.directivity,
        mask: option.mask,
        parallel: option.parallel,
    };
    solve_batched(
        backend,
        geometry,
        foci,
        wavelength,
        &setup,
        phases,
        intensities,
        |backend, g, amps, _, _| {
            let b = backend.batch_back_prop(g);
            let r = backend.batch_gemm(g, &b);
            let mut gamma = backend.batch_repeat_gemv_normalized(
                &r,
                backend.batch_gemv(&r, amps),
                amps,
                option.repeat.get() - 1,
            );
            backend.batch_amplitude_correct(&mut gamma, amps);
            backend.batch_gemv(&b, &gamma)
        },
    )
}

File: doc/codes/csharp/examples/api/compute/holo/gspat.cs
using System.Numerics;
using AUTD3;
using AUTD3.Holo;
using ControlPoint = AUTD3.Holo.ControlPoint;
using static AUTD3.Units;
using static AUTD3.Holo.HoloUnits;

namespace DocSamples.ApiComputeHoloGspat;

internal static class Sample
{
    internal static void Run()
    {
        var geometry = new Geometry(new[] { new Autd3(Vector3.Zero) });

        var center = geometry.Center + new Vector3(0.0f, 0.0f, 150.0f);
        var foci = new[]
        {
            new ControlPoint(center + new Vector3(-30.0f, 0.0f, 0.0f), 2.5e3f * Pa),
            new ControlPoint(center + new Vector3(30.0f, 0.0f, 0.0f), 2.5e3f * Pa),
        };

        var wavelength = Pattern.Wavelength(340.0f * m / s);
        uint repeat = 100;
        var constraint = EmissionConstraint.Clamp(Intensity.Min, Intensity.Max);
        var directivity = Directivity.Sphere;
        var backend = new NalgebraBackend();
        var mask = TransducerMask.AllEnabled;
        var option =
            // ANCHOR: option
            new GspatOption(
                repeat,
                constraint,
                directivity,
                backend,
                mask
            )
            // ANCHOR_END: option
            ;
        var dst = geometry.PatternBuffer();
        // ANCHOR: api
        Holo.Gspat(geometry, foci, wavelength, option, dst);
        // ANCHOR_END: api
    }
}
